include/sys/abd.h

```diff
@@ -0,0 +1,230 @@
+
+#ifndef _KERNEL
+
+#ifndef PAGE_SIZE
+#define	PAGE_SIZE 4096
+#endif
+
+#ifdef ZFS_DEBUG
+#define	DEBUG_ABD
+#endif
+
+#endif	/* !_KERNEL */
+
+typedef struct arc_buf_data {
+#ifdef DEBUG_ABD
+	char		pad[PAGE_SIZE];	/* debug, coredumps when accessed */
+#endif
+	uint32_t	abd_magic;	/* ARC_BUF_DATA_MAGIC */
+	uint32_t	abd_flags;
+	size_t		abd_size;	/* buffer size, excluding offset */
+	size_t		abd_offset;	/* offset in the first segment */
+	int		abd_nents;	/* num of sgl entries */
+	union {
+		struct scatterlist *abd_sgl;
+		void *abd_buf;
+	};
+	uint64_t __abd_sgl[0];
+} abd_t;
+
+#define	ABD_F_SCATTER	(0x0)
+#define	ABD_F_LINEAR	(0x1)
+#define	ABD_F_OWNER	(0x2)
+
+/*
+ * Convert an linear ABD to normal buffer
+ */
+#define	ABD_TO_BUF(abd)					\
+(							\
+{							\
+	ASSERT((abd)->abd_magic == ARC_BUF_DATA_MAGIC);	\
+	ASSERT_ABD_LINEAR(abd);				\
+	abd->abd_buf;					\
+}							\
+)
+
+#define	ABD_IS_SCATTER(abd)	(!((abd)->abd_flags & ABD_F_LINEAR))
+#define	ABD_IS_LINEAR(abd)	(!ABD_IS_SCATTER(abd))
+#define	ASSERT_ABD_SCATTER(abd)	ASSERT(ABD_IS_SCATTER(abd))
+#define	ASSERT_ABD_LINEAR(abd)	ASSERT(ABD_IS_LINEAR(abd))
+
+/*
+ * Allocations and deallocations
+ */
+abd_t *abd_alloc_scatter(size_t);
+abd_t *abd_alloc_linear(size_t);
+void abd_free(abd_t *, size_t);
+abd_t *abd_get_offset(abd_t *, size_t);
+abd_t *abd_get_from_buf(void *, size_t);
+void abd_put(abd_t *);
+
+/*
+ * ABD operations
+ */
+void abd_iterate_rfunc(abd_t *, size_t,
+    int (*)(const void *, uint64_t, void *), void *);
+void abd_iterate_wfunc(abd_t *, size_t,
+    int (*)(void *, uint64_t, void *), void *);
+void abd_iterate_func2(abd_t *, abd_t *, size_t, size_t,
+    int (*)(void *, void *, uint64_t, uint64_t, void *), void *);
+void abd_copy_off(abd_t *, abd_t *, size_t, size_t, size_t);
+void abd_copy_from_buf_off(abd_t *, const void *, size_t, size_t);
+void abd_copy_to_buf_off(void *, abd_t *, size_t, size_t);
+int abd_cmp(abd_t *, abd_t *, size_t);
+int abd_cmp_buf_off(abd_t *, const void *, size_t, size_t);
+void abd_zero_off(abd_t *, size_t, size_t);
+#ifdef _KERNEL
+int abd_copy_to_user_off(void __user *, abd_t *, size_t, size_t);
+int abd_copy_from_user_off(abd_t *, const void __user *, size_t, size_t);
+int abd_uiomove_off(abd_t *, size_t, enum uio_rw, uio_t *, size_t);
+int abd_uiocopy_off(abd_t *, size_t, enum uio_rw, uio_t *, size_t *,
+    size_t);
+unsigned int abd_scatter_bio_map_off(struct bio *, abd_t *, unsigned int,
+    size_t);
+unsigned long abd_bio_nr_pages_off(abd_t *, unsigned int, size_t);
+
+#define	abd_bio_map_off(bio, abd, size, off)				\
+(									\
+{									\
+	unsigned int ___ret;						\
+	if (ABD_IS_LINEAR(abd))						\
+		___ret = bio_map(bio, ABD_TO_BUF(abd) + (off), size);	\
+	else								\
+		___ret = abd_scatter_bio_map_off(bio, abd, size, off);	\
+	___ret;								\
+}									\
+)
+#endif	/* _KERNEL */
+
+/*
+ * Borrow a linear buffer for an ABD
+ * Will allocate if ABD is scatter
+ */
+#define	abd_borrow_buf(a, n)			\
```

Comment:
Is not about the performance. Is that we don't want to copy over stuff that we shouldn't.
But I guess we can add a length argument to abd_get_offset and hopefully get around this problem.